bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/action/EcobeeActions.java

```diff
@@ -0,0 +1,783 @@
+        }
+        if (heatHoldTemp != null) {
+            params.put("heatHoldTemp", heatHoldTemp);
+        }
+        if (holdClimateRef != null) {
+            params.put("holdClimateRef", holdClimateRef);
+        }
+        return setHold(params, holdType, holdHours, startDateTime, endDateTime);
+    }
+
+    public static boolean setHold(@Nullable ThingActions actions, @Nullable QuantityType<Temperature> coolHoldTemp,
+            @Nullable QuantityType<Temperature> heatHoldTemp, @Nullable String holdClimateRef,
+            @Nullable Date startDateTime, @Nullable Date endDateTime, @Nullable String holdType,
+            @Nullable Number holdHours) {
+        if (actions == null) {
+            throw new IllegalArgumentException("actions cannot be null");
+        }
+        try {
+            return (Boolean) actions.getClass()
+                    .getDeclaredMethod("setHold", QuantityType.class, QuantityType.class, String.class, Date.class,
+                            Date.class, String.class, Number.class)
+                    .invoke(actions, coolHoldTemp, heatHoldTemp, holdClimateRef, startDateTime, endDateTime, holdType,
+                            holdHours);
+        } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException
+                | InvocationTargetException e) {
+            LOGGER.info("EcobeeActions: Exception getting or invoking method", e);
+            return false;
+        }
+    }
+
+    /**
+     * Set a hold by providing a parameter map
+     */
+    @RuleAction(label = "Set Hold", description = "The set hold function sets the thermostat into a hold with the specified event parameters.")
+    public @ActionOutput(name = "success", type = "java.lang.Boolean") Boolean setHold(
+            @ActionInput(name = "params", description = "The map of hold parameters.") @Nullable Map<String, Object> params,
+            @ActionInput(name = "holdType", description = "(opt) The hold duration type. Valid values: dateTime, nextTransition, indefinite, holdHours.") @Nullable String holdType,
+            @ActionInput(name = "holdHours", description = "(opt) The number of hours to hold for, used and required if holdType='holdHours'.") @Nullable Number holdHours,
+            @ActionInput(name = "startDateTime", description = "(opt) The start date in thermostat time.") @Nullable Date startDateTime,
+            @ActionInput(name = "endDateTime", description = "(opt) The end date in thermostat time.") @Nullable Date endDateTime) {
+        LOGGER.debug("EcobeeActions: Action 'SetHold' called");
+        if (params == null) {
+            throw new IllegalArgumentException("params cannot be null");
+        }
+        EventDTO event = new EventDTO();
+        for (String key : params.keySet()) {
+            Object value = params.get(key);
+            switch (key) {
+                case "isOccupied":
+                    event.isOccupied = ((Boolean) value);
```

Comment:
Why aren't any of these casts getting checked beforehand?

Reply:
Good question. I adapted this from the OH1 version of the binding, which didn't do any checking. Seems like the correct approach would be to throw `IllegalArgumentException`. Agree?

Reply:
Isn't IllegalArguementException an unchecked exception, or is that not a concern?

Reply:
I wasn't sure, which is why I was asking.

Most of the [thing actions](https://github.com/openhab/openhab-addons/search?q=actions+illegalargumentexception&unscoped_q=actions+illegalargumentexception) already call `IllegalArgumentException` (including this one), so I was assuming it would be ok to call it in this situation as well. In addition, it will cause the rule to fail, which probably is the desired behavior if the action is passed something that it can't deal with.